src/sinks/datadog/logs.rs

```diff
@@ -260,40 +275,60 @@ impl HttpSink for DatadogLogsJsonService {
                 count: events.len(),
             });
         }
-        self.config
-            .build_request(self.uri.as_str(), "application/json", body)
+        self.config.build_request(
+            self.uri.as_str(),
+            api_key.as_str(),
+            "application/json",
+            body,
+        )
     }
 }
 
 #[async_trait::async_trait]
 impl HttpSink for DatadogLogsTextService {
-    type Input = Bytes;
-    type Output = Vec<Bytes>;
+    type Input = PartitionInnerBuffer<Bytes, String>;
+    type Output = PartitionInnerBuffer<Vec<Bytes>, String>;
+
+    fn encode_event(&self, mut event: Event) -> Option<EncodedEvent<Self::Input>> {
+        let api_key = event
+            .metadata_mut()
+            .datadog_api_key
+            .take()
+            .unwrap_or_else(|| self.config.api_key.clone());
 
-    fn encode_event(&self, event: Event) -> Option<EncodedEvent<Self::Input>> {
         encode_event(event, &self.config.encoding).map(|e| {
             emit!(DatadogLogEventProcessed {
                 byte_size: e.item.len(),
                 count: 1,
             });
-            e
+
+            EncodedEvent {
+                item: PartitionInnerBuffer::new(e.item, api_key),
+                metadata: e.metadata,
+            }
+            //}::new(PartitionInnerBuffer::new(e.item, api_key)).with_metadata(e.metadata)
         })
     }
 
-    async fn build_request(&self, events: Self::Output) -> crate::Result<http::Request<Vec<u8>>> {
+    async fn build_request(&self, events: Self::Output) -> crate::Result<Request<Vec<u8>>> {
+        let (events, api_key) = events.into_parts();
         let body: Vec<u8> = events.into_iter().flat_map(Bytes::into_iter).collect();
+
         self.config
-            .build_request(self.uri.as_str(), "text/plain", body)
+            .build_request(self.uri.as_str(), api_key.as_str(), "text/plain", body)
     }
 }
 
 /// The healthcheck is performed by sending an empty request to Datadog and checking
 /// the return.
-async fn healthcheck<T, O>(sink: T, client: HttpClient) -> crate::Result<()>
+async fn healthcheck<T, O>(sink: T, client: HttpClient, api_key: String) -> crate::Result<()>
 where
-    T: HttpSink<Output = Vec<O>>,
+    T: HttpSink<Output = PartitionInnerBuffer<Vec<O>, std::string::String>>,
 {
-    let req = sink.build_request(Vec::new()).await?.map(Body::from);
+    let req = sink
+        .build_request(PartitionInnerBuffer::new(Vec::new(), api_key.clone()))
```

Comment:
Best to call `Vec::with_capacity` instead of `::new` if there are going to be multiple `Event` instances pushed into the buffer. `Vec` will start out with no space allocated and double each item pushed in, so it's cheaper to just eat up a little memory at the outset and avoid the 0 -> 1 -> 2 -> 4 etc reallocs. 

If there aren't going to be multiple `Event` instances pushed into `PartitionInnerBuffer` we might reconsider the use of this type.